### ISSUE TITLE
Bug: game resets everytime the app is restarted

### ISSUE THREAD

**Issue by rmrt1n** [flathub/com.shatteredpixel.shatteredpixeldungeon#37]
After updating to v1.2.0, the game deletes all of my previous runs. It happens whenever I open the app, all settings and runs are reset to default. I've tried deleting all data & reinstalling but the issue still persists.

**Comment by dschier-wtd** [comment]
Had a look and the change "Improved resilience for save games" is in fact a change to an absolute path, which is currently not persisted. I will have a look if this can be adapted and provide a fix during the day.

**Comment by dschier-wtd** [comment]
I have provided a test build in the linked PR. I woulld appreciate, if you can have an additional look @rmrt1n

**Comment by rmrt1n** [comment]
I have tested the changes. Everything works as expected. Thanks!

**Comment by de-souza** [comment]
Hi, the issue is still here in the case of a fresh install. For a workaround, creating an empty directory at `~/.shatteredpixel/shattered-pixel-dungeon/` fixes the issue.

**Comment by dschier-wtd** [comment]
I will have another look. Thank you for reporting.

**Comment by de-souza** [comment]
In fact, for me adding `--persist=var` to the permissions is sufficient to solve the issue. However, there is a bug [upstream](https://github.com/00-Evan/shattered-pixel-dungeon/issues/939) related to the save path so this might change in the next version.

**Comment by dschier-wtd** [comment]
I am not sure how the "persist=var" should fix the linked issue, since it seems to be a mere copy inconsistency.

Doing a fresh installation (system and user) also results in a properly saving system, but old save games (pre-1.2.0) are not moved properly, as described [here](https://github.com/00-Evan/shattered-pixel-dungeon/issues/939).

Could you provide more details about the situation for you?

**Comment by de-souza** [comment]
My problem may be related with the fact that I am using Fedora Silverblue, and therefore my home directory is located at `/var/home/myusername/`.

Here is how I came up with `--persist=var`:

1. Make a fresh install. Launch the game, enter the dungeon, save and exit. Launch the game again. -> The last save does not exist and it looks like a fresh install.
2. Launch the game with `--filesystem=host` or `--filesystem=home`, enter the dungeon, save and exit. Launch the game again. -> The last save is found, and a new directory with the save files is created at: `/var/home/myusername/var/home/myusername/.var/app/com.shatteredpixel.shatteredpixeldungeon/data.shatteredpixel/shattered-pixel-dungeon/`.
3. Remove this new directory and launch the game with `--persist=var` instead. Everything works and now the saves files are created at `/var/home/myusername/.var/app/com.shatteredpixel.shatteredpixeldungeon/var/home/myusername/.var/app/com.shatteredpixel.shatteredpixeldungeon/data.shatteredpixel/shattered-pixel-dungeon/`.

I guess everything works directly if somebody's home directory is in `/home/myusername/` instead of `/var/home/myusername/`, because there is already `--persist=home` in the permissions. If my theory is correct, the game would then create the save files in `/home/username/.var/app/com.shatteredpixel.shatteredpixeldungeon/home/username/.var/app/com.shatteredpixel.shatteredpixeldungeon/data.shatteredpixel/shattered-pixel-dungeon/`.

I don't know whether these repetition in the save path comes from flatpak, from the game, and whether this will behave differently with version 1.2.1. I also don't know why I have to do all this, because on my system there is a symlink `/home` pointing to `/var/home`. Maybe flatpak has a problem with this? If that's the case, it could be that other apps have similar issues?

**Comment by dschier-wtd** [comment]
Thank you for the deatied explanation. I will spin up a silverblue instance over the weekend and give it a shot.

**Comment by dschier-wtd** [comment]
That sounds like a nice idea. I will have a look <3

**Comment by dschier-wtd** [comment]
hmm, seems like 1.3.0 is already in beta since yesterday. I will test, if this fixes the situation and come back afterwards with either the workaround or a statement how to proceed.

**Comment by santiagocezar** [comment]
Ah, well I thought that as the game isn't really in a working state, it could be a quick fix to, at least for now, make it playable.

**Comment by dschier-wtd** [comment]
At least for the majority it is in a working state. As far as I know, the only issue is on Silverblue, where home is a symlink to /var/home. A workaround is also provided https://github.com/flathub/com.shatteredpixel.shatteredpixeldungeon/issues/37#issuecomment-1086285902

@santiagocezar Do you see another issue?

**Comment by dschier-wtd** [comment]
I have started to work on the beta, which is available in #42 

You can test out the new build via:

```sh
flatpak install --user https://dl.flathub.org/build-repo/95942/com.shatteredpixel.shatteredpixeldungeon.flatpakref
```

**Comment by santiagocezar** [comment]
Sadly it still doesn't work

**TLDR:** the game joins `$HOME` with `$XDG_DATA_HOME`, and that's just incorrect

Trying to debug the issue, I ran the game with `strace` grepping `open` and I found this:
```
openat(AT_FDCWD, "/var/home/santi/var/home/santi/.var/app/com.shatteredpixel.shatteredpixeldungeon/data/.shatteredpixel/shattered-pixel-dungeon/settings.xml", O_WRONLY|O_CREAT|O_TRUNC, 0666 <unfinished ...>
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

An extra `/var/home/santi`?. Looking at the [base path for other platforms](https://github.com/00-Evan/shattered-pixel-dungeon/blob/master/desktop/src/main/java/com/shatteredpixel/shatteredpixeldungeon/desktop/DesktopLauncher.java#L135), I think I found the actual issue:

- Windows base path is `AppData/Roaming/.shatteredpixel/Shattered Pixel Dungeon/`
- For macOS is `Library/Application Support/Shattered Pixel Dungeon/`
- And for Linux is `$XDG_DATA_HOME`, so `/var/home/santi/.var/app/com.shatteredpixel.shatteredpixeldungeon/data` in flatpak

In Windows and macOS the path is relative, but on Linux instead is absolute. I'm guessing somewhere else in the code (maybe even in lwjgl3) it joins the home directory with the base path thus ending up with the duplicate `/var/home/santi`.

Not really sure how to proceed, I'll try to make a PR to SPD adding something to truncate `$HOME` from the path. For now, adding --persist=var might be the best option.

**Comment by dschier-wtd** [comment]
Following the initial request (persisting var) to adopt for silverblue.
As you stated, it's broken in the game and the developers seems also to be aware of it.

The new test build for the flatpak is running. Let's see.

**Comment by santiagocezar** [comment]
It's now fixed in 1.3.0-RC1! Works in master branch by just updating the game binary URL

**Comment by dschier-wtd** [comment]
I will check and update the 1.3.0 PR accordingly. Also time for some new screenshots.

**Comment by dschier-wtd** [comment]
Now 1.3.0 is on the branch and I am testing everything. Also new screenshots are in and the workaround of persistent home/var seems not longer needed.

@santiagocezar would be nice, if you can have a look and write down your feedback, before merge.

**Comment by santiagocezar** [comment]
LGTM. Tested saving a game, unlocking a character, changing settings and saving again after deleting the game data

**Comment by dschier-wtd** [comment]
merged and fixed :)
